### PR TITLE
Improve Setup Scripts

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
#### Mac:
* Asking permission from user for every installation - this includes an `all` option for eliminating subsequent questions
* Improved error handling - warning upon non-zero exit code, error if the command execution fails
* Adding `-V` option for `sudo gem install cocoapods`. This triggers a verbose output
* Outputting multiple license acceptances for android tools
* Creating ~/.cocoapods directory in case it does not exist

#### Windows:
* Asking permission from user for every installation - this includes an `all` option for eliminating subsequent questions
* Improved error handling - warning upon non-zero exit code
* Initial check for latest .NET Framework installed - should be at least 4
* Check for correct Chocolatey installation
* Fixed the syntax in docs header on invoking the setup script from the command prompt
* Added example syntax in docs header for invoking the setup script from powershell itself

#### Doctor command:
* Upon detecting problems
  * Ask user whether to open up docs
  * Ask user whether to run setup script
* Both questions have **NO** as default for non-interactive consoles and **YES** in interactive

Addresses some parts of https://github.com/NativeScript/NativeScript/issues/1556

Ping @rosen-vladimirov and @teobugslayer

### CODE REVIEW

File: setup/native-script.ps1 [hunk 1]
```diff
@@ -1,30 +1,76 @@
 
 # A PowerShell script to set up Windows machine for NativeScript development
-# To run it against the PRODUCTION branch (only one supported with self-elevation) use
-# @powershell -NoProfile -ExecutionPolicy Bypass -Command "iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.ps1'))" 
+# NOTE: The scripts requires at least a version 4.0 .NET framework installed
+# To run it inside a COMMAND PROMPT against the production branch (only one supported with self-elevation) use
+# @powershell -NoProfile -ExecutionPolicy Bypass -Command "iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.ps1'))"
+# To run it inside a WINDOWS POWERSHELL console against the production branch (only one supported with self-elevation) use
+# start-process -FilePath PowerShell.exe -NoNewWindow -Wait -ArgumentList "-NoProfile -ExecutionPolicy Bypass -Command iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.ps1'))"
```
Comment (teobugslayer) [inline setup/native-script.ps1]: Well, you are inside a powershell, why are you running another one? What about simply executing:

```
iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.ps1'))
```
?

File: setup/native-script.ps1 [hunk 2]
```diff
@@ -1,30 +1,76 @@
...
+	start-process -FilePath PowerShell.exe -Verb Runas -Wait -ArgumentList "-NoProfile -ExecutionPolicy Bypass -Command iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.ps1'))"
+	exit 0
+}
+
+# Help with installing other dependencies
+$script:answer = ""
+function Install($programName, $message, $script, $shouldExit) {
+	if ($script:answer -ne "all") {
+		Write-Host -BackgroundColor Black -ForegroundColor Green "Allow the script to install $($programName)?"
+		Write-Host -BackgroundColor Black "Tip: Note that if you type all you won't be prompted for subsequent installations"
```
Comment (teobugslayer) [inline setup/native-script.ps1]: What happens when the foreground color is black too? is the text visible?

File: setup/native-script.ps1 [hunk 3]
```diff
@@ -1,30 +1,76 @@
...
+		do {
+			$script:answer = (Read-Host "y/n/all").ToLower()
+		} until ($script:answer -eq "y" -or $script:answer -eq "n" -or $script:answer -eq "all")
```
Comment (teobugslayer) [inline setup/native-script.ps1]: Why `all` and not `a`?

File: setup/native-script.ps1 [hunk 4]
```diff
@@ -1,30 +1,76 @@
...
+}
+
+# Help with installing other dependencies
+$script:answer = ""
+function Install($programName, $message, $script, $shouldExit) {
+	if ($script:answer -ne "all") {
+		Write-Host -BackgroundColor Black -ForegroundColor Green "Allow the script to install $($programName)?"
+		Write-Host -BackgroundColor Black "Tip: Note that if you type all you won't be prompted for subsequent installations"
+		do {
+			$script:answer = (Read-Host "y/n/all").ToLower()
```
Comment (teobugslayer) [inline setup/native-script.ps1]: I'd write the text like `(Y)es/(N)o/(A)ll` or something similar to both highlight the hotkey and convey its meaning.

File: setup/native-script.ps1 [hunk 5]
```diff
@@ -1,30 +1,76 @@
...
+		} until ($script:answer -eq "y" -or $script:answer -eq "n" -or $script:answer -eq "all")
+
+		if ($script:answer -eq "n") {
+			Write-Host -BackgroundColor Black -ForegroundColor Yellow "You've chosen not to install $($programName) some features of NativeScript may not work correctly if you haven't already installed it"
```
Comment (teobugslayer) [inline setup/native-script.ps1]: Nitpicking. Using contractions like `you've` is discouraged in formal writing.

File: setup/native-script.ps1 [hunk 6]
```diff
@@ -1,30 +1,76 @@
...
+			return
+		}
+	}
+
+	Write-Host -BackgroundColor Black $message
```
Comment (teobugslayer) [inline setup/native-script.ps1]: background is set but not the foreground. what is the result?

File: setup/native-script.rb [hunk 7]
```diff
@@ -1,19 +1,23 @@
...
-# To run it against PRODUCTION branch (recommended) use
-# ruby -e "$(curl -fsSL https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.rb)"
-# To run it against MASTER branch (usually only developers of NativeScript need to) use
-# ruby -e "$(curl -fsSL https://raw.githubusercontent.com/NativeScript/nativescript-cli/master/setup/native-script.rb)"
+# To run it against PRODUCTION branch (only one supported with self-elevation) use
+# sudo ruby -e "$(curl -fsSL https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.rb)"
 
 # Only the user can manually download and install Xcode from App Store
+unless Process.uid == 0
+  # Self-elevate
```
Comment (teobugslayer) [inline setup/native-script.rb]: You can elevate only on Windows. Perhaps "execute as root"

File: setup/native-script.rb [hunk 8]
```diff
@@ -23,32 +27,73 @@
...
+    puts "Allow the script to install " + program_name + "?"
+    if show_all_option
+      puts "Note that if you type all you won't be prompted for subsequent installations"
+    end
+
+    loop do
+      puts show_all_option ? "y/n/all" : "y/n"
+      $answer = gets.chomp.downcase
+      is_answer_yn = $answer == "y" || $answer == "n"
+      break if show_all_option ? is_answer_yn || $answer == "all" : is_answer_yn
```
Comment (teobugslayer) [inline setup/native-script.rb]: Again, why y/n but all? Very inconsistent.

File: setup/native-script.rb [hunk 9]
```diff
@@ -23,32 +27,73 @@
...
-system('echo "export ANDROID_HOME=/usr/local/opt/android-sdk" >> ~/.profile')
+# the -p flag is set in order to ensure zero status code even if the directory exists
+execute("mkdir -p ~/.cocoapods", "There was a problem in creating ~/.cocoapods directory")
+install("CocoaPods", "Installing CocoaPods... This might take some time, please, be patient.", 'gem install cocoapods -V', true)
 
 puts "Configuring your system for Android development... This might take some time, please, be patient."
-system "echo yes | /usr/local/opt/android-sdk/tools/android update sdk --filter tools,platform-tools,android-23,build-tools-23.0.2,extra-android-m2repository --all --no-ui"
+# Note that multiple license acceptances may be required, hence the multiple y answers
+# the android tool will introduce a --accept-license option in subsequent releases
+execute("(for i in {1..5}; do echo y; sleep 4; done) | /usr/local/opt/android-sdk/tools/android update sdk --filter tools,platform-tools,android-23,build-tools-23.0.2,extra-android-m2repository --all --no-ui",
```
Comment (teobugslayer) [inline setup/native-script.rb]: Have you verified that installing all dependencies at once work on a clean machine? I used to have problems with this scenario.

File: setup/native-script.ps1 (hunk 2 above)
Comment (Mitko-Kerezov) [inline setup/native-script.ps1]: According to my tests the default Foreground color is `white` - so there are no problems at all

File: setup/native-script.ps1 (hunk 6 above)
Comment (Mitko-Kerezov) [inline setup/native-script.ps1]: As mentioned above, the default value for foreground color seems to be white

File: setup/native-script.ps1 (hunk 2 above)
Comment (Mitko-Kerezov) [inline setup/native-script.ps1]: In fact it turns out the default Background color is `black`, so I'll remove that too

**Comment by vchimev** [comment]
run ci

File: setup/native-script.rb (hunk 9 above)
Comment (Mitko-Kerezov) [inline setup/native-script.rb]: I haven't encountered any issues regarding the installation of all the dependencies (apart from accepting licenses in this weird fashion)

**Comment by Mitko-Kerezov** [comment]
Ping @teobugslayer with fixed comments

File: setup/native-script.ps1 [hunk 10]
```diff
@@ -1,30 +1,76 @@
...
+# @powershell -NoProfile -ExecutionPolicy Bypass -Command "iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.ps1'))"
+# To run it inside a WINDOWS POWERSHELL console against the production branch (only one supported with self-elevation) use
+# iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/NativeScript/nativescript-cli/production/setup/native-script.ps1'))
+
+# Check if latest .NET framework installed is at least 4
+$dotNetVersions = Get-ChildItem 'HKLM:\SOFTWARE\Microsoft\NET Framework Setup\NDP' -recurse | Get-ItemProperty -name Version,Release -EA 0 | Where { $_.PSChildName -match '^(?!S)\p{L}'} | Select Version
+$latestDotNetVersion = $dotNetVersions.GetEnumerator() | Sort-Object Version | Select-Object -Last 1
+$latestDotNetMajorNumber = $latestDotNetVersion.Version.Split(".")[0]
+if ($latestDotNetMajorNumber -lt 4) {
+	Write-Host -ForegroundColor Red "Running this script requires at least a version 4.0 .NET framework installed"
```
Comment (teobugslayer) [inline setup/native-script.ps1]: From Iva: "To run this script, you need .NET 4.0 or later installed"

File: setup/native-script.ps1 [hunk 11]
```diff
@@ -1,30 +1,76 @@
...
+	if ((Read-Host "Do you want to open Microsoft's Download Center in your default browser(y/n)") -eq 'y') {
```
Comment (teobugslayer) [inline setup/native-script.ps1]: From Iva: "Do you want to open Microsoft Download Center (y/n)"

**Comment by teobugslayer** [comment]
Fix the messages, squash the commits and merge it! :+1: and excellent work!

**Pull request merged**